internal/controller/controllers/cluster_controller.go

```diff
@@ -270,6 +271,43 @@ func (r *ClusterReconciler) notifyPullSecretUpdate(ctx context.Context, isPullSe
 	return nil
 }
 
+func (r *ClusterReconciler) deregisterClusterIfNeeded(ctx context.Context, key types.NamespacedName) (ctrl.Result, error) {
+
+	buildReply := func(err error) (ctrl.Result, error) {
+		reply := ctrl.Result{}
+		if err == nil {
+			return reply, nil
+		}
+		reply.RequeueAfter = defaultRequeueAfterOnError
+		err = errors.Wrapf(err, "failed to deregister cluster: %s", key.Name)
+		r.Log.Error(err)
+		return reply, err
+	}
+
+	c, err := r.Installer.GetClusterByKubeKey(key)
+
+	if gorm.IsRecordNotFoundError(err) {
+		// return if from any reason cluster is already deleted from db (or never existed)
+		return buildReply(nil)
+	}
+
+	// todo check error type, retry for 5xx fail on 4xx
+	if err != nil {
+		return buildReply(err)
+	}
+
+	r.Log.Infof("Cluster resource deleted, deregister cluster: %s", c.ID.String())
+
+	err = r.Installer.DeregisterClusterInternal(ctx, installer.DeregisterClusterParams{
```

Comment:
`IsHTTPError(err, http.StatusNotFound)` is not relevant because you already called for `r.Installer.GetClusterByKubeKey(key)` 
```suggestion
	err = r.Installer.DeregisterClusterInternal(ctx, installer.DeregisterClusterParams{
		ClusterID: *c.ID,
	})
	if err != nil{
		return buildReply(err)
	}
	
	r.Log.Infof("Cluster resource deleted, deregister cluster: %s", c.ID.String())
	return buildReply(nil)
```